Tests/NIOTests/ChannelTests.swift

```diff
@@ -1442,4 +1442,88 @@ public class ChannelTests: XCTestCase {
         try clientChannel.closeFuture.wait()
         try serverChannel.syncCloseAcceptingAlreadyClosed()
     }
+
+    func testWeDontJamSocketsInANoIOState() throws {
+        final class ReadDelayer: ChannelDuplexHandler {
+            typealias InboundIn = Any
+            typealias InboundOut = Any
+            typealias OutboundIn = Any
+            typealias OutboundOut = Any
+
+            private var ctx: ChannelHandlerContext!
+            public var reads = 0
+            private var readCountPromise: EventLoopPromise<Void>!
+            private var waitingForReadPromise: EventLoopPromise<Void>?
+
+            func handlerAdded(ctx: ChannelHandlerContext) {
+                self.ctx = ctx
+                self.readCountPromise = ctx.eventLoop.newPromise()
+            }
+
+            public func expectRead(loop: EventLoop) -> EventLoopFuture<Void> {
+                return loop.submit {
+                    self.waitingForReadPromise = loop.newPromise()
+                }.then { (_: Void) in
+                    return self.waitingForReadPromise!.futureResult
+                }
+            }
+
+            func channelReadComplete(ctx: ChannelHandlerContext) {
+                self.waitingForReadPromise?.succeed(result: ())
+                self.waitingForReadPromise = nil
+            }
+
+            func read(ctx: ChannelHandlerContext) {
+                self.reads += 1
+
+                // Allow the first read through.
+                if self.reads == 1 {
+                    self.ctx.read()
+                }
+            }
+
+            public func issueDelayedRead() {
+                self.ctx.read()
+            }
+        }
+
+        let group = MultiThreadedEventLoopGroup(numThreads: 1)
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+        let readDelayer = ReadDelayer()
+
+        let serverChannel = try ServerBootstrap(group: group)
+            .serverChannelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+            .childChannelInitializer {
+                $0.pipeline.add(handler: readDelayer)
+            }
+            .bind(host: "127.0.0.1", port: 0).wait()
+
+        let clientChannel = try ClientBootstrap(group: group)
+            .connect(to: serverChannel.localAddress!).wait()
+
+        // We send a first write and expect it to arrive.
+        var buffer = clientChannel.allocator.buffer(capacity: 12)
+        let firstReadPromise = readDelayer.expectRead(loop: serverChannel.eventLoop)
```

Comment:
nit: move this under the "assert of the writeAndFlush" as it is not needed before.

Reply:
Yes it is: if it's done after it's possible that the server will get the `channelReadComplete` before this call makes it onto the event loop.